LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions

OrdiNeu opened the pull request:
Removed two items from the sidebar to make room for this.

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms
as answers to questions
Move vocabulary to props so that it can be edited externally

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as
answers to questions
Add required ancestors

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Add portal to move selection element to different DOMs.
Add test rig to test selector.

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Add a maximum number of selected components

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Prevent the user from adding duplicates

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Allow the input to become disabled when there is no input

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Add support for default suggestions

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Prevent default suggestions from being re-loaded when the portal moves

marta- reviewed (changes requested):
A couple of issues with the suggestions popup:

1/ In narrow windows the (i) goes on the next line:
![image](https://user-images.githubusercontent.com/651980/63712959-ba2a2e00-c80c-11e9-9002-26822dde74ef.png)

In short windows, the popup jumps at the top of the window when scrolling on it:
![image](https://user-images.githubusercontent.com/651980/63713248-43416500-c80d-11e9-8804-74873efb319e.png)

3/ When clicking on info buttons, the corresponding suggestion should not be added. instead, suggestions popup should be kept open, and the info modal should be displayed on top.

4/ The "close" button of the info modal is too pow, should be closer to the top right corner. Additionally, it should not become white on hover and it should be a [multiplication sign](https://www.fileformat.info/info/unicode/char/00d7/index.htm), not letter X.

![image](https://user-images.githubusercontent.com/651980/63713541-dd091200-c80d-11e9-946c-60b112d35b9a.png)

5/ Resizing the window while the info modal is displayed pushes the modal all the way to the left:
![image](https://user-images.githubusercontent.com/651980/63713622-03c74880-c80e-11e9-89c0-a1e648347787.png)

6/ In the info modal, the distance between the subtitle (e.g. "Is a type of") and the paragrapgh below it is larger than the distance to the paragraph above it, which makes the whole layout look messy. This seems to be caused by line breaks generated between the subtitle and the normal text. Suggested spacing (please also note the new character and positioning for the close button and the font size for the vocabulary name):

![image](https://user-images.githubusercontent.com/651980/63716305-a8985480-c813-11e9-8f73-8e6ef41b15e9.png)

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Move colour and theming over to an exported indexStyle from
MaterialDashboardReact

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Prevent selection state from activating when the user clicks the info
button. Also move checkbox labels to a parent FormControlInput

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Use a maximum of "0" for the default selection results

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Improve style of info box and documentation

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Handle the disabled/enabled state of the text input by showing an error
message or clearing the input when disabled, as appropriate

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Fix popper positioning and close button.
Remove unused default props in selectChild

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Rename VocabularyChild to VocabularyEntry

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Stylistic improvements, console error removal

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Move the arguments of MakeChildFindingRequest to an object

marta- commented on modules/thesaurus/src/main/frontend/src/query/util.jsx:
```diff
@@ -33,10 +33,31 @@ export function MakeRequest(URL, callback) {
+    // Start with the suggest URL
+    var URL = `${REST_URL}/${vocabulary}/suggest?`;
+
+    // Add default options
```
There's a more elegant (in my opinion) and easier to maintain way of building your options from defaults + overrides:

1/ Declare the defaults: 
```
const DEFAULT_REQ_OPTIONS = {
  sort: "name asc",
  ...
};
```

2/ Initialize local options with the defaults
```
var crtRequestOpt = {};
Object.extend(crtRequestOpt, DEFAULT_REQ_OPTIONS);
```

3/ Overwrite the defaults with any options given in the parameters:
```
Object.extend(crtRequestOpt, requestOpt)
```

marta- commented on modules/thesaurus/src/main/frontend/src/query/util.jsx:
```diff
@@ -33,10 +33,31 @@ export function MakeRequest(URL, callback) {
+        requestOpt['maxResults'] = '10000';
+    }
+    if (!requestOpt.hasOwnProperty('customFilter')) {
+        var escapedId = requestOpt['input'].replace(":", "\\:"); // URI Escape the : from HP: for SolR
```
```suggestion
        var escapedId = requestOpt['input'].replace(":", "\\:"); // URI Escape the : from vocabulary term identifier (e.g.HP:#######) for SolR
```

marta- commented on modules/thesaurus/src/main/frontend/src/selector/select.jsx:
```diff
@@ -0,0 +1,219 @@
+
+      // Determine the name from our vocab
+      var testId = id;
+      var escapedId = id.replace(":", "\\:"); // URI Escape the : from HP: for SolR
```
```suggestion
      var escapedId = id.replace(":", "\\:"); // URI Escape the : from vocabulary term identifiers (e.g.HP:#######) 
```

marta- commented on modules/thesaurus/src/main/frontend/src/selector/select.jsx:
```diff
@@ -0,0 +1,219 @@
+
+VocabularySelector.defaultProps = {
+    name: "VocabularySelector",
+    source: "hpo",
```
IMO there should be no default for this property. Specifying the source should be mandatory, or else there will be no suggestions.

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
More generic comment by @marta-

Co-Authored-By: Marta Girdea <marta.girdea@gmail.com>

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Code cleanup.
Remove |-s and retheme vocab browser and infobox.
Fix arrow key event not focusing the menu
Fix console error regarding <div>s being a child of <p>

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Reformat close button

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Handle radio buttons when the max number of selectable options is 1

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Add more options to the test rig

marta- commented on modules/material-dashboard/src/main/frontend/src/MaterialDashboardReact/ExternalThemes/queryStyle.jsx:
```diff
@@ -132,10 +134,13 @@ const thesaurusStyle = theme => ({
   },
   // The following ensures poppers are placed above the presentation, which has zIndex 1300
```
Deprecated info in this comment?

marta- reviewed (approved):
Looks good.

Only minor improvements to be done **in another iteration**:
* better layout of the vocabulary browser on narrow screen
* ~when focusing an "Other" input, select any text that is in it~ (done)
* consider using styling instead of `&nbsp;`s and `<br/>`s for layout

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Select text on text focus

marta- reviewed (approved):
Approving the latest changes

OrdiNeu committed:
LFS-111: As a user, I can look up and select standard vocabulary terms as answers to questions
Allow the Vocabulary searcher to input children.
Fix styles.

sdumitriu commented on modules/thesaurus/src/main/frontend/src/query/query.jsx:
```diff
@@ -156,31 +199,31 @@ class Thesaurus extends React.Component {
+                        <a
+                          href="http://human-phenotype-ontology.github.io/"
+                          target="_blank"
+                          title="The Human Phenotype Ontology project: linking molecular biology and disease through phenotype data. Sebastian Köhler, Sandra C Doelken, Christopher J. Mungall, Sebastian Bauer, Helen V. Firth, et al. Nucl. Acids Res. (1 January 2014) 42 (D1): D966-D974 doi:10.1093/nar/gkt1026. Current version: releases/2018-10-09"
```
This will have to be taken from the vocabulary itself.

sdumitriu commented on modules/thesaurus/src/main/frontend/src/selector/selectTestRig.jsx:
```diff
@@ -0,0 +1,89 @@
+// @material-ui/core
+import VocabularySelector from "./select.jsx";
+
+class VocabularyTester extends React.Component {
```
This doesn't look like production code. Should it be removed, or moved as a test source?
